streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -17,242 +17,225 @@
-        }
+    public InternalTopicManager(StreamsConfig config, int replicationFactor) {
+        this.config = config;
+        this.streamsKafkaClient = new DefaultKafkaClientSupplier().getStreamKafkaClient(config);
+        this.kafkaClient = streamsKafkaClient.getKafkaClient();
+        this.replicationFactor = replicationFactor;
     }
 
-    private List<Integer> getBrokers() {
-        List<Integer> brokers = new ArrayList<>();
-        for (String broker: zkClient.getChildren(ZK_BROKER_PATH)) {
-            brokers.add(Integer.parseInt(broker));
-        }
-        Collections.sort(brokers);
-
-        log.debug("Read brokers {} from ZK in partition assignor.", brokers);
 
-        return brokers;
+    public void shutDown() throws IOException {
+        this.streamsKafkaClient.shutdown();
     }
+    public void makeReady(String topic, int numPartitions, boolean compactTopic) {
 
-    @SuppressWarnings("unchecked")
-    private Map<Integer, List<Integer>> getTopicMetadata(String topic) {
-        String data = zkClient.readData(ZK_TOPIC_PATH + "/" + topic, true);
-
-        if (data == null) return null;
+        Node brokerNode = kafkaClient.leastLoadedNode(new SystemTime().milliseconds());
 
-        try {
-            ObjectMapper mapper = new ObjectMapper();
+        // Delete the topic if it exists before.
+        boolean topicExists = topicExists(topic, brokerNode);
+        if (topicExists) {
+            deleteTopic(topic, brokerNode);
+            log.debug("Deleted: " + topic);
+        }
+        // Create a new topic
+        createTopic(topic, numPartitions, replicationFactor, compactTopic, brokerNode);
+        log.debug("Created: " + topic);
+    }
 
-            Map<String, Object> dataMap = mapper.readValue(data, new TypeReference<Map<String, Object>>() {
 
-            });
+    /**
+     * Check if the topic exists already
+     *
+     * @param topic
+     * @param brokerNode
+     * @return
+     */
+    private boolean topicExists(String topic, Node brokerNode) {
+        MetadataRequest metadataRequest = new MetadataRequest(Arrays.asList(topic));
+        String brokerId = Integer.toString(brokerNode.id());
+        RequestSend send = new RequestSend(brokerId,
+                kafkaClient.nextRequestHeader(ApiKeys.METADATA),
+                metadataRequest.toStruct());
 
-            Map<Integer, List<Integer>> partitions = (Map<Integer, List<Integer>>) dataMap.get("partitions");
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
 
-            log.debug("Read partitions {} for topic {} from ZK in partition assignor.", partitions, topic);
+            }
+        };
+
+        // Send the async request to check the topic existance
+        ClientRequest clientRequest = new ClientRequest(new SystemTime().milliseconds(), true, send, callback);
+        SystemTime systemTime = new SystemTime();
+        int iterationCount = 0;
```

Comment:
iterationCount is never incremented so the loop below is essentially `while(true)`